### PR TITLE
feat: get or create a team conversation

### PR DESCRIPTION
**Pull request by trOnk12**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

We need to able to create a one to one conversation with a team mate, for that I created a 

### Solutions
createOne2OneConversationWithTeamMate(...) and getOne2OneConversationByUserId(...) inside ConversationRepository

I also created CreateOneToOneConversationUseCase to combine the buisness logic inside 

Note. 

Please <3 review it carefully as I do not have that much context I may be wrong on some points, so plx show me da way ! 💟 

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:177]: Repository should insert the new conversation into the DB after creating it.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/CreateOneToOneConversationUseCase.kt
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/CreateOneToOneConversationUseCase.kt:10]:
```suggestion
class GetOrCreateOneToOneConversationUseCase(
```

It's a bit clearer what it does.

We can call it multiple times without being afraid of creating multiple 1:1 conversations. Also, if some app actually wants to create multiple conversations, the name would clearly indicate why it doesn't work.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:188]: This function name says it would return a whole Conversation, but it only returns the `ConversationId?`.

Perhaps we can return everything, and whoever is calling this function can use the ID or whatever else they need.
```suggestion
    override suspend fun getOne2OneConversationDetailsByUserId(otherUserId: UserId): Either<StorageFailure, ConversationDetails.OneOne?> {
        return wrapStorageRequest {
            observeConversationList()
                .flatMapMerge { it.asFlow() }
                .flatMapMerge { getConversationDetailsById(it.id) }
                .filterIsInstance<ConversationDetails.OneOne>()
                .firstOrNull { otherUserId == it.otherUser.id }
        }
    }
```

Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:188]: Could you add some tests for this function?

**Review by vitorhugods — CHANGES_REQUESTED** [review]
Nice and clean work!
Just needs to persist the conversation after creating it.

A couple of other smaller suggestions.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:177]: This is overlapping with my PR we can reuse things

Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:177]: You should we able to delegate to the  `createGroupConversation()` method I added in https://github.com/wireapp/kalium/pull/366/files#diff-7cc6befc172b3bde25bc0dae25c48a8c3b54476ec722bd1be49d9b82359693df

Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:177]: Since you are creating a team one-to-one you should assign the `convTeamInfo` field though.

**Review by typfel — CHANGES_REQUESTED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:177]: @typfel awesome, waiting for your merge than :)

Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:177]: > @typfel awesome, waiting for your merge than :)

Just waiting for approvals ;)

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:46]:
```suggestion
    suspend fun getOneToOneConversationDetailsByUserId(otherUserId: UserId): Either<CoreFailure, ConversationDetails.OneOne?>
```

Let's be consist on the OneToOne naming?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:215]: So we observe the entire conversation list and pick out the right conversation. Any reason why we don't fetch the one-to-one conversation directly?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/GetOrCreateOneToOneConversationUseCase.kt
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/GetOrCreateOneToOneConversationUseCase.kt:33]: You only want the conversationId and not he whole conversation object?

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/conversation/CreateConversationRequest.kt
Comment (typfel) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/conversation/CreateConversationRequest.kt:16]:
```suggestion
    // the name is optional in swagger but this should not be the case
    // since there is endpoint for one2one and self conversations
```

Let's delete this comment now that name is optional.

**Review by typfel — CHANGES_REQUESTED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/GetOrCreateOneToOneConversationUseCase.kt
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/GetOrCreateOneToOneConversationUseCase.kt:33]: I think for now we can try to work with ConversationId, or do you see any case that we would need the whole object using it in AR ?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:215]: discussed on PM that we will optimize it later on

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/GetOrCreateOneToOneConversationUseCase.kt
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/GetOrCreateOneToOneConversationUseCase.kt]: Only need an ID for now I'm fine with that. Not hard to change.

**Review by typfel — APPROVED** [review]

**Pull request merged**
